src/mpid/ch4/netmod/ofi/ofi_win.h

```diff
@@ -52,20 +58,60 @@ static inline int MPIDI_OFI_win_progress_fence(MPIR_Win * win)
 
     r = MPIDI_OFI_WIN(win).syncQ;
 
-    while (r) {
-        MPIDI_OFI_win_request_t *next = r->next;
-        MPIDI_OFI_rma_done_event(NULL, (MPIR_Request *) r);
-        r = next;
+    /* Should not free the RMA request when do_free = 0; manual RMA flush could happen in
+     * the middle of issuing one MPI-level RMA op and it could lead to premature freeing of
+     * the request. */
+    if (do_free) {
+        while (r) {
+            MPIDI_OFI_win_request_t *next = r->next;
+            MPIDI_OFI_rma_done_event(NULL, (MPIR_Request *) r);
+            r = next;
+        }
+
+        MPIDI_OFI_WIN(win).syncQ = NULL;
+        MPIDI_OFI_WIN(win).progress_counter = 1;
     }
 
-    MPIDI_OFI_WIN(win).syncQ = NULL;
   fn_exit:
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_OFI_WIN_PROGRESS_FENCE);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_OFI_WIN_PROGRESS_FENCE_IMPL);
     return mpi_errno;
   fn_fail:
     goto fn_exit;
 }
 
+/* If the OFI provider does not have automatic progress, check to see if the progress engine should
+ * be manually triggered to keep performance from suffering when doing large, non-continuguous
+ * transfers. */
+static inline int MPIDI_OFI_win_trigger_rma_progress(MPIR_Win * win)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    if (!MPIDI_OFI_ENABLE_DATA_AUTO_PROGRESS && MPIR_CVAR_CH4_OFI_RMA_PROGRESS_INTERVAL != -1) {
+        if (MPIDI_OFI_WIN(win).progress_counter % MPIR_CVAR_CH4_OFI_RMA_PROGRESS_INTERVAL == 0) {
+            MPIDI_OFI_win_progress_fence_impl(win, false);
+            MPIDI_OFI_WIN(win).progress_counter = 1;
+        }
+        MPIDI_OFI_WIN(win).progress_counter++;
+    }
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_OFI_WIN_COUNTER_CHECK);
```

Comment:
Sorry, I just noticed this EXIT macro doesn't match the function name, nor does it have a DECL or ENTER counterpart.

Reply:
@wesbland can you fix it?

Reply:
Done. 👍

Reply:
Arg...I didn't check first and overwrote your changes. Sorry.

Reply:
OK, I'll take care of it.